### PR TITLE
Mail/Msg refactoring

### PR DESCRIPTION
**Pull request by TornaxO7**
Hi! I did some changes to the `Msg` struct. You should be able now to use it for everything which is related to Msgs/Mails. This means:

- You can build templates with it
- You can reply with it
- Edit the body
- Add new attachments

and so on.

I'd really appreaciate it, if this could be merged, because the "new" `Msg` struct will be able to store all related information of a Msg which should make it easier for the TUI to display them.

# Hint
It's not completely done yet. You can see below in the `TODO` section what is missing, but the rest should work.

# TODO
- [X] Finish the Documentation
- [X] Finish/Add Tests

### CODE REVIEW

**Comment by TornaxO7** [comment]
Alright, you can start reviewing this now @soywod .

**Comment by soywod** [comment]
Since it is a big PR, I will fetch your branch and check that everything works as expected (especially the vim plugin). I will try to do it next week, I let you know. Thank you!

**Comment by TornaxO7** [comment]
No problem! Take your time.

**Comment by TornaxO7** [comment]
I also recommend to read through the `cargo doc --open` which I provided. It should help to understand it.

**Comment by TornaxO7** [comment]
Ok so looks like, that I have to fix some stuff first again. Sorry for the complications!

File: src/msg/cli.rs
Comment (soywod) [inline src/msg/cli.rs:664]: I don't think it is necessary to have the type here and with the collect

**Review by soywod — CHANGES_REQUESTED** [review]
I started to check a good part of your PR (mostly the syntax / style). Here the 2 remarks I have that applies to all the PR:

- Try to avoid comment separators (see my first review comment)
- Try not to change logs style (you added a capital letter), and style in general (for example using different name for the Msg ressource like Message, Mail, Email etc)

You did a giant job! Well done :+1: 

I will check the rest of the code + try the branch in an other review.

File: src/flag/model.rs
Comment (TornaxO7) [reply src/flag/model.rs:14]: This is the version: `rustfmt 1.4.37-nightly (7c3872e 2021-06-24)`

And this are my settings:

<details>

```toml
color = "Always"
enum_discrim_align_threshold = 100
format_code_in_doc_comments = true
# format_strings = true
match_block_trailing_comma = false
# max_width = 80
# # reorder_impl_items = true
struct_field_align_threshold = 100
wrap_comments = true
```

</details>

File: src/flag/model.rs
Comment (TornaxO7) [reply src/flag/model.rs:54]: That's interesting! I'll take a look into this comment!

Comment (TornaxO7) [reply src/flag/model.rs:54]: Ok, so I finished it now. Hm... I like to write stuff in a row or group them. Like putting all traits in a row or all "helper functions". Is that considered as a bad practice as well? Because it's in my opinion nice to see it like that if you open the file:

image

Or is it not even allowed to use `// -- <section> --` ? :thinking:

File: src/msg/cli.rs
Comment (TornaxO7) [reply src/msg/cli.rs:261]: I thought that it would look nicer since it represents a starting line and looks like this: `...] List command matched`. But I can revert that

File: src/msg/cli.rs
Comment (TornaxO7) [reply src/msg/cli.rs:877]: Yes you're right, looks like I forgot to change some names.

File: src/msg/cli.rs
Comment (TornaxO7) [reply src/msg/cli.rs:584]: When I saw it for the first time, I didn't knew what I got. So it helped for me to know the data type here. But if you don't want to have it, then I can remove it. (But I'd like to let it like that).

**Comment by TornaxO7** [comment]
Alright, should be done now!

**Comment by TornaxO7** [comment]
@fancycade if soywod would merge this branch, does your issue with the message-id persist? Could you please try [my branch](https://github.com/TornaxO7/himalaya/tree/mail_refactor)?

**Comment by TornaxO7** [comment]
> I started to check a good part of your PR (mostly the syntax / style). 

Nice! :D

> Here the 2 remarks I have that applies to all the PR:
> - Try to avoid comment separators (see my first review comment)
> - Try not to change logs style (you added a capital letter), and style in general (for example using different name for the Msg 
                ressource like Message, Mail, Email etc)
These should be fixed now.

> You did a giant job! Well done +1

Thank you for these kind words, but please review it completely first before it might turn into the other side :eyes: 

> I will check the rest of the code + try the branch in an other review.

:+1:

**Comment by TornaxO7** [comment]
@soywod in my latest commit (Named: `Idea(?)`) I renamed some functions so the modules/cli.rs are gonna have two main-functions: `matches` and `subcmds` without the prefix. What do you think about that? In my opinion it's almost the same, the "advantage" (again: In my opinion) is that you can say in the documentation, that each module should include at least two functions: `matches` and `subcmds` or something like that.

Should I rather reverse this change or is it fine for you?

**Comment by fancycade** [comment]
@TornaxO7 Same issue as before. Complains about missing Message-Id header.

**Comment by TornaxO7** [comment]
@fancycade my latest commit should fix that now, can you try it again please?

File: src/msg/model.rs
Comment (fancycade) [inline src/msg/model.rs:623]: @TornaxO7 You probably don't want this anymore since it is overriding the code above.

Comment (TornaxO7) [reply src/msg/model.rs:623]: Oops, you are right, forgot to remove it. Thank you for the hint!

**Comment by TornaxO7** [comment]
@fancycade should be done now!

**Comment by fancycade** [comment]
@TornaxO7 Tried testing but I couldn't build your branch:

```
    Updating git repository `https://github.com/TornaxO7/lettre`
error: failed to get `lettre` as a dependency of package `himalaya v0.4.0 (/home/fc/himalaya)`

Caused by:
  failed to load source for dependency `lettre`

Caused by:
  Unable to update https://github.com/TornaxO7/lettre?branch=content_type_serde#3fbc8d74

Caused by:
  object not found - no match for id (3fbc8d74766111f82698103dde4f71a8f3550b8f); class=Odb (9); code=NotFound (-3)
```

**Comment by TornaxO7** [comment]
What the, I'll try to fix that in the afternoon.

**Comment by TornaxO7** [comment]
@fancycade could you please try it again now?

**Comment by fancycade** [comment]
@TornaxO7 Seems you broke the prompt for write (only thing I've tested).

himalaya write goes straight to:
```
(s)end, (e)dit, (l)ocal/(r)emote draft or (d)iscard?
```

Without opening the editor.

Upon pressing edit I can open the editor, but when I go to select send it just loops back to the prompt.

All I'm doing to test the Message-Id header issue is sending a test email to
~fancycade/public-inbox@lists.sr.ht. If it is successful it will show up on the list.
If not you will get back a reply that message-id isn't present.

Feel free to send test messages there, I haven't used it for anything else.

**Comment by TornaxO7** [comment]
> himalaya write goes straight to:

> (s)end, (e)dit, (l)ocal/(r)emote draft or (d)iscard?
> Without opening the editor.

Hm... well, I thought that this would be fine, but going straight to the editing is probably better, you're right.

> Upon pressing edit I can open the editor, but when I go to select send it just loops back to the prompt.

You get stuck in this "loop" because something happened during the convertion into a sendable message. I changed it now, so you'll get a message what gone wrong. Could you please try it again @fancycade ? I'm sorry, that you have to try it out that often...

**Comment by fancycade** [comment]
> You get stuck in this "loop" because something happened during the convertion into a sendable message.

Yep, that was the issue. Showing the error message is certainly helpful.

```
[Error] Invalid email domain
Header-Field-Name: 'To'
The word which let this error occur: '~fancycade/public-inbox@lists.sr.ht '
```

Looks like what was happening was that there was a space at the end of the domain. This seems to be "naturally" happening, as in I'm not adding a space at the end. So maybe there is an extra space in the template, but a better solution is to probably strip the To header as I'm sure user's will get frustrated dealing with whitespace errors.

Message-Id did work though!
https://lists.sr.ht/~fancycade/public-inbox/97bb365b-75b7-4d78-8ca8-9ec8c2cd75a0%40mycanofbeans.com

**Comment by TornaxO7** [comment]
That's starnge, might I ask you how you invoke my pull request? Something like that?

```bash
cargo r -- template new -t tornax07@gmail.com -B "open source is cool"  | cargo r -- send
```
Because this works on my machine. Well, I think that I can fix this by trimming the mail-addresses.

**Comment by fancycade** [comment]
```
cargo build --release
./target/release/himalaya write
```

**Comment by TornaxO7** [comment]
@fancycade I can't really reproduce it. Could you please give me the steps?

**Comment by TornaxO7** [comment]
Because I don't really think, that this:
```rust
        // add "to"
        for mailaddress in &self.envelope.to {
            msg = msg.to(match mailaddress.trim().parse() {
                Ok(to) => to,
                Err(err) => {
                    return Err(
                        ErrorKind::Header(err.to_string(), "To", mailaddress.to_string()).into(),
                    )
                }
            });
        }
```
is gonna fixing this problem.

**Comment by TornaxO7** [comment]
@fancycade (sorry if I interrupt you, just a reminder)

**Comment by fancycade** [comment]
@TornaxO7 

I'm just doing a basic write:

```
himalaya write
```

I have my default editor set to emacs. Some editors have subtle differences with whitespacing. The issue shows up on the To field.

**Comment by TornaxO7** [comment]
@fancycade bug should be fixed now

**Comment by fancycade** [comment]
@TornaxO7 :+1: Tried it out and it seems better now.

**Comment by TornaxO7** [comment]
Thank you for testing! 👍

**Comment by TornaxO7** [comment]
Soywod? Did you forgot this pull request or do you just have not much time at the moment? :eyes:

**Comment by soywod** [comment]
@TornaxO7 don't worry, I don't forget you :smile: I just can't get free time for myself to check it. I will try this week!

**Comment by TornaxO7** [comment]
Oh ok, I don't want to put you under pressure! It should be just a reminder.

**Comment by soywod** [comment]
PS: you seem to have a conflict on `msg/model.rs`, it may be because of the `Message-Id` (see https://github.com/soywod/himalaya/pull/171).

**Comment by TornaxO7** [comment]
***Nani???*** How could the test fail!?

That's my output if I run `cargo test`:

<details>

```
running 13 tests
test msg::envelope::tests::test_convert_cow_u8_to_string ... ok
test msg::envelope::tests::test_convert_vec_address_to_string ... ok
test msg::model::tests::test_new_with_envelope ... ok
test msg::model::tests::test_edit_body ... ok
test msg::model::tests::test_new ... ok
test table::tests::max_shrink_width ... ok
test table::tests::row_smaller_than_head ... ok
test table::tests::row_bigger_than_head ... ok
test msg::envelope::tests::test_merge_addresses_to_one_line ... ok
test msg::model::tests::test_change_to_forwarding ... ok
test table::tests::basic_shrink ... ok
test msg::model::tests::test_parse_from_str ... ok
test msg::model::tests::test_change_to_reply ... ok

test result: ok. 13 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.02s

     Running unittests (target/debug/deps/himalaya-969fbe6e89243804)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/imap_test.rs (target/debug/deps/imap_test-bb1a7fdb8895a523)

running 2 tests
test mbox ... ok
test msg ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.61s

   Doc-tests himalaya

running 19 tests
test src/msg/attachment.rs - msg::attachment::Attachment (line 101) - compile ... ok
test src/msg/attachment.rs - msg::attachment::Attachment (line 126) - compile ... ok
test src/msg/envelope.rs - msg::envelope::Envelope (line 185) - compile ... ok
test src/msg/body.rs - msg::body::Body (line 86) ... ok
test src/config/model.rs - config::model::Account::get_full_address (line 118) ... ok
test src/msg/attachment.rs - msg::attachment::Attachment::new (line 40) ... ok
test src/config/model.rs - config::model::Account::get_full_address (line 139) ... ok
test src/msg/attachment.rs - msg::attachment::Attachment::from_parsed_mail (line 67) ... ok
test src/msg/body.rs - msg::body::Body (line 101) ... ok
test src/msg/model.rs - msg::model::Msg::edit_body (line 374) - compile ... ok
test src/msg/model.rs - msg::model::Msg::edit_body (line 357) - compile ... ok
test src/msg/envelope.rs - msg::envelope::Envelope (line 35) ... ok
test src/msg/body.rs - msg::body::Body::get_content (line 34) ... ok
test src/msg/model.rs - msg::model::Msg::to_sendable_msg (line 488) - compile ... ok
test src/msg/envelope.rs - msg::envelope::Envelope (line 368) ... ok
test src/msg/model.rs - msg::model::Msg::add_attachment (line 462) ... ok
test src/msg/envelope.rs - msg::envelope::Envelope::get_header_as_string (line 80) ... ok
test src/msg/model.rs - msg::model::Msg::new (line 111) ... ok
test src/msg/model.rs - msg::model::Msg::parse_from_str (line 419) ... ok

test result: ok. 19 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 10.53
s
```

</details>

**Comment by TornaxO7** [comment]
Oh, hold on, that's probably because there's no server for the test to try it out.

**Comment by soywod** [comment]
Yes, see https://github.com/soywod/himalaya/issues/145. To spaw a test server, check how does the CI: https://github.com/soywod/himalaya/blob/fadebf09977df5d5e571d1d9a2f73ccea44390a7/.github/workflows/tests.yaml#L15.

**Comment by TornaxO7** [comment]
@soywod I already prepared the mail-server on my machine and as you can see above, my tests are passing. o.O

**Comment by soywod** [comment]
Ow. This is strange :thinking:. I will try the branch to see if tests pass on my machine!

**Comment by TornaxO7** [comment]
https://user-images.githubusercontent.com/50843046/126969174-a62e3b93-b4e6-4416-b149-7685366afb15.mp4

Here's a demo

**Comment by TornaxO7** [comment]
I just saw, that I didn't implement the `content-transfer-encoding` like in your code [here](https://github.com/soywod/himalaya/blob/master/src/msg/model.rs#L320). Do we need it thought?

**Comment by soywod** [comment]
> I just saw, that I didn't implement the `content-transfer-encoding` like in your code [here](https://github.com/soywod/himalaya/blob/master/src/msg/model.rs#L320). Do we need it thought?

Yes, it is better to implement it also to avoid regressions.

**Comment by TornaxO7** [comment]
Should be done now.

**Review by soywod — CHANGES_REQUESTED** [review]
I tried your branch, here is my feedback (It may be incomplete, I will update it over time):

## General

- [ ] The `From` header contains only the email address (it should contains the name if available `My Name <my@email.com>`)
- [ ] The signature is missing
- [ ] The `Subject` header is missing

## Read

- [ ] The JSON output does not have the same shape (it misses `hasAttachment`)
- [ ] The JSON output should show the `text/html` version of the email if no `text/plain` is found

## Reply & Forward

- [ ] The body does not contain the previous message

## Template

- [ ] The `raw` should be excluded from the JSON export ([`#[serde(skip_serializing)]`](https://serde.rs/attr-skip-serializing.html))
- [ ] The JSON export should be in `camelCase`, not in `snake_case` ([`#[serde(rename_all = "camelCase")]`](https://serde.rs/field-attrs.html))

---

In resume, try to fit as close as possible to the existing behaviour. For now, the gap is too big between how the CLI works now and how you branch works. Feel free to comment of course.

**Comment by TornaxO7** [comment]
That's strange, the points in your general section should work as expected. When is the subject missing?

I'd prefer that you create some pull requests on my fork, so I can find the diffs faster (or can I see the changes gast as well, if you apply your changes here? 🤔 Correct me, if I'm wrong).

**Comment by soywod** [comment]
> That's strange, the points in your general section should work as expected. When is the subject missing?

I tried on new message and reply if I remember well.

> I'd prefer that you create some pull requests on my fork, so I can find the diffs faster (or can I see the changes gast as well, if you apply your changes here?  Correct me, if I'm wrong).

Yes, it makes more sense. I will open PRs on your repo!

**Comment by TornaxO7** [comment]
@soywod Are you really sure that the reply didn't work? My tests have passed so it should be fine. I'll look, if my my latest commit is online. (Sorry for this late message)

**Comment by TornaxO7** [comment]
Ok, so the latest commit was always there. Anyway, I could reproduce the issue that the signature was missing in the reply. Could you please describe what you're trying to fix or start the appropriate pull request so I know, what I can fix in the meantime?

**Comment by TornaxO7** [comment]
Ok, so I quickly wrote two little fixed. More details are in the description of those two commits.

**Comment by TornaxO7** [comment]
@soywod ?

**Comment by soywod** [comment]
@TornaxO7 I think I will start by resolving conflicts (to include last master's features). I also plan to fix the signature. I see that you check if the signature exists on the account side, but it should be checked on the config side. The account signature should be taken if exists, otherwise it should fallback to the general config one. There is already a `signature()` function on the config struct that does this check. I will send you a PR on your repo when I can!

**Comment by soywod** [comment]
> When forwarding a message, himalaya, put the signature in the end of the
mail/msg. Now it's added above the '-------Forwarded Message---------' line.

Why having changed this behaviour? What are your motivations?

**Comment by TornaxO7** [comment]
> Why having changed this behaviour? What are your motivations?

Oh yes, eh I should've added an example with an explanation for this, sorry.

Suppose anyone wrote you a message:
```
From: Bla
Subject: Blub

Text goes brrrrrrrrrr

Friend Signature
```

And now you want to forward it, I thought that this:
```
From: Him
Subject: Fwd: Blub
...

My text goes brrrrrrrrr as well

My signature

----- Fowarded Message -----
Text goes brrrrrrrrrrrrrrr

Friends Signature
```

Would be better than this:
```


From: Him
Subject: Fwd: Blub
...

My text goes brrrrrrrrr as well

My signature

----- Fowarded Message -----
Text goes brrrrrrrrrrrrrrr

Friends Signature

My siganture
```

to separate the message which is fowarded with your added content.

**Comment by TornaxO7** [comment]
I hope that it was fine to read my code and hopefully easy to [understand]/[get the context]. Feel free to point out any suggestions if a documentation is missing or something is hard to read or bad written.

**Comment by TornaxO7** [comment]
> I also plan to fix the signature. I see that you check if the signature exists on the account side, but it should be checked on the config side.

I don't really understand what you mean with this. Aren't you pointing [here](https://github.com/soywod/himalaya/blob/master/src/main.rs#L83) to the selected account in the `Ctx.account` struct? So why should I still use `Ctx.config` to access some information about an account if I can already access the through `Ctx.account`?

**Comment by soywod** [comment]
> I don't really understand what you mean with this

Imagine you have this config:

```toml
name = "Clément DOUIN"
signature = "Regards,\nClément DOUIN"

[account-soywod]
name = "Soywod"

[account-anon]
name = "Anonymous"
signature = "Regards"
```

If I send an email with `account-anon`, I expect my signature to be `"Regards"`. But if I send an email with the `account-soywod` account, I expect my signature to be `"Regards,\nClément DOUIN"`. There is a main signature by default (config level) that can be overridden by accounts (account level). Because of how the model is implemented (a config contains a map of accounts), it is better to call the `signature` method from the config rather than the account.

**Comment by TornaxO7** [comment]
I'd like to speed up the process of this pull request. Could you please tell me what I should do? @soywod

**Comment by soywod** [comment]
@TornaxO7 I finally merged master into your branch and fixed conflicts. I tried to push on another branch to create a PR on your repo but I was not able (permission denied). Here the diff with your previous commit: https://github.com/soywod/himalaya/pull/173/commits/a79c3a210fd1f566fc670c93402149cedfb9523a. Now I will test and fix the tiny issues I found last time.

**Comment by TornaxO7** [comment]
> @TornaxO7 I finally merged master into your branch and fixed conflicts.

Nice, thank you!

> I tried to push on another branch to create a PR on your repo but I was not able (permission denied).

Why don't you fork my fork and create a pull request on my fork or am I misunderstanding you?

**Comment by soywod** [comment]
> Why don't you fork my fork and create a pull request on my fork or am I misunderstanding you?

Ow. I did not know it was possible to fork a fork! Good to know. I guess it's too late now that I already pushed on your branch... should I continue like this for the rest or I do the fork?

**Comment by TornaxO7** [comment]
I would appreciate it if you could create a fork of my fork and create a pull request with it, because then we can discuss it better before merging it (in my opinion).

**Comment by soywod** [comment]
I just tried to fork yours but it's not possible. It says I already have a fork and it redirects me to soywod/himalaya :disappointed:

**Comment by TornaxO7** [comment]
Is it possible that you clone my branch, create another branch, apply your fixes and create a pull request with it?

**Comment by soywod** [comment]
It's what I said in my previous message, I don't have access to create a branch on your repo.

**Comment by TornaxO7** [comment]
I've sent you an invitation to be a co-maintainer of my fork.

**Comment by soywod** [comment]
Thanks for the invitation. I revert my last commit on your branch `mail_refactor`. I was able to push my branch `mail_refactor_merge_master` and open a PR: https://github.com/TornaxO7/himalaya/pull/1.

**Comment by TornaxO7** [comment]
I don't understand why the nix-build and the imap-test fails...

**Comment by soywod** [comment]
> I don't understand why the nix-build and the imap-test fails...

About nix-build: `error: Cannot find Git revision '912ec605c5a24bbec1b1a5126c7768d80677f29a' in ref 'master' of repository 'https://github.com/TornaxO7/lettre/'! Please make sure that the rev exists on the ref you've specified or add allRefs = true; to fetchGit.`. It looks like the commit does not exist on master anymore (maybe due to a rebase?).

For the tests: did you try locally? Does it work?

**Comment by TornaxO7** [comment]
I'm getting the following error message:
```
thread 'mbox' panicked at 'called `Result::unwrap()` on an `Err` value: Error(Msg("Could not connect to IMAP server"), State { next_error: Some(Io(Custom { kind: Uncategorized, error: "failed to lookup address information: Name or service not known" })), backtrace: InternalBacktrace { backtrace: None } })', tests/imap_test.rs:15:54
```
looks like that the greenmail doesn't work? The same error is in the tests build.

**Comment by soywod** [comment]
> looks like that the greenmail doesn't work?

We are more prone to break stuff than GreenMail, so I guess it comes from us :laughing: I will try to check tomorrow.

**Comment by TornaxO7** [comment]
```java
Exception in thread "imaps:0.0.0.0:3993<-/172.17.0.1:38046" java.lang.IllegalStateException: Can not handle IMAP connection
	at com.icegreen.greenmail.imap.ImapHandler.run(ImapHandler.java:75)
	at com.icegreen.greenmail.server.AbstractServer.lambda$handleClientSocket$0(AbstractServer.java:160)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at com.icegreen.greenmail.imap.ImapHostManagerImpl.getQualifiedMailboxName(ImapHostManagerImpl.java:267)
	at com.icegreen.greenmail.imap.ImapHostManagerImpl.listMailboxes(ImapHostManagerImpl.java:219)
	at com.icegreen.greenmail.imap.ImapHostManagerImpl.listMailboxes(ImapHostManagerImpl.java:204)
	at com.icegreen.greenmail.imap.commands.ListCommand.doList(ListCommand.java:142)
	at com.icegreen.greenmail.imap.commands.ListCommand.doProcess(ListCommand.java:103)
	at com.icegreen.greenmail.imap.commands.CommandTemplate.process(CommandTemplate.java:57)
	at com.icegreen.greenmail.imap.ImapRequestHandler.doProcessRequest(ImapRequestHandler.java:98)
	at com.icegreen.greenmail.imap.ImapRequestHandler.handleRequest(ImapRequestHandler.java:53)
	at com.icegreen.greenmail.imap.ImapHandler.run(ImapHandler.java:71)
	... 2 more
```
I think my machine drunk too much coffee...

**Comment by TornaxO7** [comment]
Wohoo! Ready to merge + 100 commits! I'd say: Perfect landing :D

**Comment by TornaxO7** [comment]
Now let's see if the tests are gonna fail here....

**Comment by TornaxO7** [comment]
Oh wait, about the nix problem. How should we solve it?

**Comment by TornaxO7** [comment]
I don't understand. All tests are passing on my machine. How does the imap-test fail???

**Comment by TornaxO7** [comment]
Alright, let's fix that ~tomorrow~ today, but in the evening.

**Comment by TornaxO7** [comment]
Alright, now it should work.

**Comment by TornaxO7** [comment]
@soywod could you please approve the workflows to see if they work? The tests are working on my machine.

**Comment by TornaxO7** [comment]
oh man... at least the nix-build works. I can try fixing the workflow again this evening.

**Comment by TornaxO7** [comment]
@soywod workflow approval pls.

**Comment by TornaxO7** [comment]
@soywod workflow approval pls. I'm sorry, that I'm pinging you so often... just notify me if it annoys you.

File: .github/workflows/tests.yaml
Comment (soywod) [inline .github/workflows/tests.yaml:15]: Why did you change this? The `--rm` is for docker to auto remove the container after the tests, and the `-d` is for the container to run as daemon (in background). The `-t` and `-i` are not needed because we are not in a TTY mode (the error is quite explicit btw: `the input device is not a TTY`).

Comment (TornaxO7) [reply .github/workflows/tests.yaml:15]: oh, I just copied the stuff from my execution file. I'll revert this

Comment (soywod) [reply .github/workflows/tests.yaml:15]: It may solve the issue in fact!

**Comment by TornaxO7** [comment]
So I copied + pasted the test command from [here](https://github.com/jonhoo/rust-imap#running-the-test-suite).

**Comment by TornaxO7** [comment]
@soywod you can approve again.

**Comment by soywod** [comment]
> So I copied + pasted the test command from [here](https://github.com/jonhoo/rust-imap#running-the-test-suite).

Yes, but this is for a local usage, not in a CI (it needs to be adapted).

**Comment by TornaxO7** [comment]
o.O how do I have to change it then?

**Comment by soywod** [comment]
> o.O how do I have to change it then?

But I don't understand why you want to change this? It works on master, why don't you keep the same config?

**Comment by TornaxO7** [comment]
> But I don't understand why you want to change this? It works on master, why don't you keep the same config?

Because it worked for me, when I added this `docker pull greenmail/standalone:1.6.3` to my execution file. Before that it didn't work for me.

**Comment by soywod** [comment]
> > But I don't understand why you want to change this? It works on master, why don't you keep the same config?
> 
> Because it worked for me, when I added this `docker pull greenmail/standalone:1.6.3` to my execution file. Before that it didn't work for me.

Could you please revert the config file and see the error?

**Comment by TornaxO7** [comment]
> Could you please revert the config file and see the error?

Of course! Done! ^^

**Comment by soywod** [comment]
> Of course! Done! ^^

No, I mean revert from master. Could you replace the conf with the master's one?

**Comment by TornaxO7** [comment]
Now?

**Comment by soywod** [comment]
It's better! If you run them locally you get the same error?

**Comment by TornaxO7** [comment]
Yes
```

     Running tests/imap_test.rs (target/debug/deps/imap_test-5819bbb29558aa41)

running 2 tests
test mbox ... ok
test msg ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.83s```

**Comment by TornaxO7** [comment]
could you please do a test as well on your machine?

**Comment by soywod** [comment]
> could you please do a test as well on your machine?

They also work locally from my computer... I don't understand then :sob:

**Comment by TornaxO7** [comment]
lmao xD hm... does it need a `cargo update`?

**Comment by TornaxO7** [comment]
So one of my friends (https://github.com/Erik1000) tried `cargo test` as well and got 0 errors as well.

**Comment by soywod** [comment]
Seems to work well :tada: thank your friend for me!

**Comment by TornaxO7** [comment]
Yeah, finally! Thanks @Erik1000 !!

**Comment by Erik1000** [comment]
today i learned

**Comment by TornaxO7** [comment]
Wohooo

**Comment by TornaxO7** [comment]
all passed!

**Comment by TornaxO7** [comment]
@soywod should we move the checkout in the other workflow as well to the top?

**Comment by TornaxO7** [comment]
@soywod ? :eyes:

**Comment by soywod** [comment]
Yes do

**Comment by TornaxO7** [comment]
hold on, the nix.yaml has the checkout first already.

**Comment by TornaxO7** [comment]
So this PR should be ready to merge now :+1:

**Comment by TornaxO7** [comment]
Eh, what are you waiting for? Are you doing a review at the moment? Are you unhappy with something? (Just wondering)

**Comment by soywod** [comment]
> Eh, what are you waiting for? Are you doing a review at the moment? Are you unhappy with something? (Just wondering)

Just sleeping...

**Comment by TornaxO7** [comment]
Oh sorry 😅

**Comment by TornaxO7** [comment]
Are you still reviewing?

**Comment by soywod** [comment]
Yes, it's a big PR, I want to take time to test it well. I'm quite busy at work those days, I will finish it as soon as I can.

**Comment by TornaxO7** [comment]
Ok, I was just wondering because about five days passed now. Take your time! I don't want to put you under pressure!

**Comment by soywod** [comment]
I opened 4 issues on your repo https://github.com/TornaxO7/himalaya/issues.

**Comment by soywod** [comment]
If all tests pass, I'm ready to merge!

**Comment by TornaxO7** [comment]
What the? Why are the tests failing again? ._. They are working locally on my machine ._.

**Comment by TornaxO7** [comment]
Wait, it's still testing? o.O Nevermind my previous message :sweat_smile:

**Comment by TornaxO7** [comment]
Yehaaaw! Tests passed!

**Comment by soywod** [comment]
Congratulations :tada: and thank you for your patience!

**Comment by TornaxO7** [comment]
Thanks and thank you for accepting my PR :)

**Pull request merged**
